### PR TITLE
Test data read and write for subcycling

### PR DESCRIPTION
**Pull request by BenjaminRodenberg**
## Main changes of this PR

Adds a test to ensure that data is written and read as currently specified for subcycling. This means:

* write data functions can write arbitrary data for any time step within the window.
* read data functions will only receive for each time step data that was written by the other solver at the end of the window. 
* All data written inside the window is lost and cannot be accessed by the reading solver.

## Motivation and additional information

Currently the tests succeed in #1029 and associated PRs ~~, even though the subcycling functionality is changed and preCICE does not behave anymore as described above (see [this test](https://github.com/BenjaminRodenberg/precice/blob/d3cb5fe98c3d13f70a431a67ccf3f147956585c1/src/precice/tests/SerialTests.cpp#L434-L570)). I would rather expect a failing test here.~~. *edit:* The tests still succeed in #1029. This is due to the implementation of `readScalarData`. If no `dt` is given, it always samples at the end of the window, which is exactly the functionality described above. I still think it's reasonable to test this, because it ensures that the read functions are implemented correctly.

## Author's checklist

* [ ] I added a changelog file with this PR number in `docs/changelog/` if there are noteworthy changes.
* [x] I ran `tools/formatting/check-format` and everything is formatted correctly.
* [x] I sticked to C++14 features.
* [x] I sticked to CMake version 3.10.
* [x] I squashed / am about to squash all commits that should be seen as one.

## Reviewers' checklist

* [ ] Does the changelog entry make sense? Is it formatted correctly?
* [ ] Do you understand the code changes?
* [ ] (more questions/tasks)

### CODE REVIEW

File: src/precice/tests/SerialTests.cpp [hunk 1]
```diff
@@ -288,6 +288,142 @@ BOOST_AUTO_TEST_CASE(testExplicitWithSubcycling)
...
+
+  precice.initializeData();
+
+  while (precice.isCouplingOngoing()) {
+    double readTime;
+    if (context.isNamed("SolverOne")) {
+      readTime = timewindow * windowDt; // solver one lags one window behind solver two.
+    } else {
+      readTime = (timewindow + 1) * windowDt;
+    }
```
[the remark this answers is not in the document]
Comment (BenjaminRodenberg) [reply src/precice/tests/SerialTests.cpp:391]: We are doing explicit coupling here. Does this explain the lag or still unclear? Maybe I'm also missing something.

File: src/precice/tests/SerialTests.cpp [hunk 2]
```diff
@@ -288,6 +288,142 @@ BOOST_AUTO_TEST_CASE(testExplicitWithSubcycling)
...
+    writeDataID   = precice.getDataID("DataTwo", meshID);
+    writeFunction = dataTwoFunction;
+    BOOST_TEST(writeDataID == 3);
+    readDataID   = precice.getDataID("DataOne", meshID);
+    readFunction = dataOneFunction;
+    BOOST_TEST(readDataID == 2);
+  }
+
+  writeFunction(1, 1);
+  readFunction(1, 1);
```
[the remark this answers is not in the document]
Comment (BenjaminRodenberg) [reply src/precice/tests/SerialTests.cpp:336]: I'll try to remove this. Maybe just sloppy on my side. Maybe something related to the lambda functions above. I will try to find out.

File: src/precice/tests/SerialTests.cpp [hunk 3]
```diff
@@ -288,6 +288,142 @@ BOOST_AUTO_TEST_CASE(testExplicitWithSubcycling)
...
+  vertexIDs[0] = precice.setMeshVertex(meshID, Eigen::Vector3d(0.0, 0.0, 0.0).data());
+
+  int    nSubsteps  = 4; // perform subcycling on solvers. 4 steps happen in each window.
+  int    nWindows   = 5; // perform 5 windows.
+  double maxDt      = precice.initialize();
+  double windowDt   = maxDt;
+  int    timestep   = 0;
+  int    timewindow = 0;
+  double dt         = windowDt / nSubsteps; // Timestep length desired by solver. E.g. 4 steps  with size 1/4
+  dt += windowDt / nSubsteps / nSubsteps;   // increase timestep such that we get a non-matching subcycling. E.g. 3 step with size 5/16 and 1 step with size 1/16.
```
[the remark this answers is not in the document]
Comment (BenjaminRodenberg) [reply src/precice/tests/SerialTests.cpp:354]: Good point. Something like this ok?

```
double dt = windowDt / (nSubsteps - 0.5);
```

Then we get for `nSubsteps = 4` 4 steps of size (2/7, 2/7, 2/7, 1/7).

File: src/precice/tests/SerialTests.cpp [hunk 4]
```diff
@@ -288,6 +288,142 @@ BOOST_AUTO_TEST_CASE(testExplicitWithSubcycling)
...
+    BOOST_TEST(readDataID == 1);
+  } else {
+    BOOST_TEST(context.isNamed("SolverTwo"));
+    meshID = precice.getMeshID("MeshTwo");
+    BOOST_TEST(meshID == 1);
+    writeDataID = precice.getDataID("DataTwo", meshID);
+    BOOST_TEST(writeDataID == 3);
+    readDataID = precice.getDataID("DataOne", meshID);
+    BOOST_TEST(readDataID == 2);
+  }
```
[the remark this answers is not in the document]
Comment (BenjaminRodenberg) [reply src/precice/tests/SerialTests.cpp:373]: Yes. I will remove this.

File: src/precice/tests/SerialTests.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (uekerman) [reply src/precice/tests/SerialTests.cpp:354]: Better already. In my opinion, you could also hard-code variables in tests. Readability (= debugability :smile) goes over flexibility.

File: src/precice/tests/SerialTests.cpp [hunk 5]
```diff
@@ -288,6 +288,142 @@ BOOST_AUTO_TEST_CASE(testExplicitWithSubcycling)
...
+  } else {
+    BOOST_TEST(context.isNamed("SolverTwo"));
+    meshID = precice.getMeshID("MeshTwo");
+    BOOST_TEST(meshID == 1);
+    writeDataID   = precice.getDataID("DataTwo", meshID);
+    writeFunction = dataTwoFunction;
+    BOOST_TEST(writeDataID == 3);
+    readDataID   = precice.getDataID("DataOne", meshID);
+    readFunction = dataOneFunction;
+    BOOST_TEST(readDataID == 2);
```
[the remark this answers is not in the document]
Comment (BenjaminRodenberg) [reply src/precice/tests/SerialTests.cpp]: Removed in 8cb883ecc31659a2303bd83b39ba4ceb05644081

File: src/precice/tests/SerialTests.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (BenjaminRodenberg) [reply src/precice/tests/SerialTests.cpp]: Removed in 8cb883ecc31659a2303bd83b39ba4ceb05644081

File: src/precice/tests/SerialTests.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (BenjaminRodenberg) [reply src/precice/tests/SerialTests.cpp]: Done in 8cb883ecc31659a2303bd83b39ba4ceb05644081

File: src/precice/tests/SerialTests.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (BenjaminRodenberg) [reply src/precice/tests/SerialTests.cpp]: I pushed the change above in 615536871d1910e4ca6b48a8d4126a42d86634bd

**Review by uekerman — APPROVED** [review]

**Pull request merged**